src/corefx/System.Globalization.Native/pal_calendarData.c

```diff
@@ -351,58 +290,77 @@ EnumSymbols
-                 CalendarId calendarId,
-                 UDateFormatSymbolType type,
-                 int32_t startIndex,
-                 EnumCalendarInfoCallback callback,
-                 const void* context)
+static int32_t EnumSymbols(const char* locale,
+                           CalendarId calendarId,
+                           UDateFormatSymbolType type,
+                           int32_t startIndex,
+                           EnumCalendarInfoCallback callback,
+                           const void* context)
 {
     UErrorCode err = U_ZERO_ERROR;
-    UDateFormat* pFormat = udat_open(UDAT_DEFAULT, UDAT_DEFAULT, locale, nullptr, 0, nullptr, 0, &err);
-    UDateFormatHolder formatHolder(pFormat, err);
+    UDateFormat* pFormat = udat_open(UDAT_DEFAULT, UDAT_DEFAULT, locale, NULL, 0, NULL, 0, &err);
 
     if (U_FAILURE(err))
-        return false;
+        return FALSE;
 
     char localeWithCalendarName[ULOC_FULLNAME_CAPACITY];
     strncpy(localeWithCalendarName, locale, ULOC_FULLNAME_CAPACITY);
     uloc_setKeywordValue("calendar", GetCalendarName(calendarId), localeWithCalendarName, ULOC_FULLNAME_CAPACITY, &err);
 
-    if (U_FAILURE(err))
-        return false;
-
-    UCalendar* pCalendar = ucal_open(nullptr, 0, localeWithCalendarName, UCAL_DEFAULT, &err);
-    UCalendarHolder calendarHolder(pCalendar, err);
+    UCalendar* pCalendar = ucal_open(NULL, 0, localeWithCalendarName, UCAL_DEFAULT, &err);
 
     if (U_FAILURE(err))
-        return false;
+    {
+        udat_close(pFormat);
+        return FALSE;
+    }
 
     udat_setCalendar(pFormat, pCalendar);
 
     int32_t symbolCount = udat_countSymbols(pFormat, type);
+    UChar stackSymbolBuf[100];
+    UChar* symbolBuf;
 
-    for (int32_t i = startIndex; i < symbolCount; i++)
+    for (int32_t i = startIndex; U_SUCCESS(err) && i < symbolCount; i++)
     {
         UErrorCode ignore = U_ZERO_ERROR;
-        int symbolLen = udat_getSymbols(pFormat, type, i, nullptr, 0, &ignore);
-
-        std::vector<UChar> symbolBuf(symbolLen + 1, '\0');
+        int symbolLen = udat_getSymbols(pFormat, type, i, NULL, 0, &ignore) + 1;
 
-        udat_getSymbols(pFormat, type, i, symbolBuf.data(), symbolBuf.size(), &err);
+        if (symbolLen <= sizeof(stackSymbolBuf) / sizeof(stackSymbolBuf[0]))
+        {
+            symbolBuf = stackSymbolBuf;
+        }
+        else
+        {
+            symbolBuf = calloc(symbolLen, sizeof(UChar));
+            if (symbolBuf == NULL)
+            {
+                err = U_MEMORY_ALLOCATION_ERROR;
+                break;
+            }
+        }
 
-        assert(U_SUCCESS(err));
+        udat_getSymbols(pFormat, type, i, symbolBuf, symbolLen, &err);
 
-        if (U_FAILURE(err))
-            return false;
+        if (U_SUCCESS(err))
+        {
+            callback(symbolBuf, context);
+        }
 
-        callback(symbolBuf.data(), context);
+        if (symbolBuf != stackSymbolBuf)
+        {
+            free(symbolBuf);
```

Comment:
>free [](start = 12, length = 4)

would it be better to use the allocated buffer through the loop and then free only once after the loop done or when you need to reallocate?

this is just suggestion but you don't need to to do it if you think this will complicate the code especially I am expecting most of the symbols will fit on the stack buffer

Reply:
Most of the things really fit into the stack buffer, so I don't expect the allocation path to be hit. As such it didn't make much sense to optimize it.